cylc/flow/tui/updater.py

```diff
@@ -0,0 +1,311 @@
+#!/usr/bin/env python3
+# THIS FILE IS PART OF THE CYLC WORKFLOW ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Contains the logic for updating the Tui app."""
+
+from asyncio import (
+    run,
+    sleep,
+    gather,
+)
+from copy import deepcopy
+from multiprocessing import Queue
+from time import time
+
+from zmq.error import ZMQError
+
+from cylc.flow.exceptions import (
+    ClientError,
+    ClientTimeout,
+    CylcError,
+    WorkflowStopped,
+)
+from cylc.flow.id import Tokens
+from cylc.flow.network.client_factory import get_client
+from cylc.flow.network.scan import (
+    filter_name,
+    graphql_query,
+    is_active,
+    scan,
+)
+from cylc.flow.task_state import (
+    TASK_STATUSES_ORDERED,
+)
+from cylc.flow.tui.data import (
+    QUERY
+)
+from cylc.flow.tui.util import (
+    compute_tree,
+    suppress_logging,
+)
+from cylc.flow.workflow_status import (
+    WorkflowStatus,
+)
+
+
+def get_default_filters():
+    """Return default task/workflow filters.
+
+    These filters show everything.
+    """
+    return {
+        'tasks': {
+            # filtered task statuses
+            state: True
+            for state in TASK_STATUSES_ORDERED
+        },
+        'workflows': {
+            # filtered workflow statuses
+            **{
+                state.value: True
+                for state in WorkflowStatus
+            },
+            # filtered workflow ids
+            'id': '.*',
+        }
+    }
+
+
+class Updater():
+    """The bit of Tui which provides the data.
+
+    It lists workflows using the "scan" interface, and provides detail using
+    the "GraphQL" interface.
+
+    """
+
+    # the maximum time to wait for a workflow update
+    CLIENT_TIMEOUT = 2
```

Comment:
If the workflow takes more than 2 seconds to send its response, then it will time out. Tui will put a message under the workflow saying that it has timed out, but will retry with the next update cycle.

Previously both the application and updater used the same update period of 1 second. If it took longer to get the response and do the required processing, then Tui would crash.